Documentation Improvements (#4368)
cocotb/cocotb: docs/source/coroutines.rst

and resumes after one of the Triggers or Tasks fires.
It returns the result of ``await``\ ing the Task or Trigger that fired first.
Below we see it used to implement a timeout.

.. code-block:: python3

    @cocotb.test
    async def test_quiesce_or_timeout(dut):

        # generate stimulus and drive it to the design
        for trans in generate_transactions():
            await drive(dut.intf, trans)

        # wait for the design to quiesce or timeout
        timeout = Timer(10, "us")
        result = await First(timeout, quiesce())
        assert result is not timeout, "Design has hung!"

Fortunately for users timeouts are a common operation and cocotb provides :func:`~cocotb.triggers.with_timeout`.
The second section in the above code using it would be ``await with_timeout(quiesce(), 10, "us")``.

.. note::

    :class:`~cocotb.triggers.First` does *not* cancel Tasks that did not complete after it returns.
    This means that Tasks passed to it are *still running*.
    You may need to cancel those Tasks with :meth:`.Task.cancel`.

Determining Which Task Finishes First
-------------------------------------

> :class:`~cocotb.triggers.First` can be used to determine which of multiple Tasks complete first using the following idiom.

Reviewer: Might link the word "complete" to the API description.

    Removed ``cocotb.fork``.

Waiting For Multiple Events Simultaneously
==========================================

> Occassionally you'll need to wait for either one of many Tasks or Triggers to fire,

Reviewer: ```suggestion
Occasionally you'll need to wait for either one of many Tasks or Triggers to fire,
```

or a collection of Tasks or Triggers to fire.
This is what :class:`~cocotb.triggers.First` and :class:`~cocotb.triggers.Combine` provide, respectively.

> :class:`~cocotb.triggers.First` is like ``await``\ ing multiple Triggers or Tasks at the same time,

Reviewer: ```suggestion
Determining Which Event Happened First
--------------------------------------

:class:`~cocotb.triggers.First` is like ``await``\ ing multiple Triggers or Tasks at the same time,
```

and resumes after one of the Triggers or Tasks fires.

> It returns the result of ``await``\ ing the Task or Trigger the fired first.

Reviewer: ```suggestion
It returns the result of ``await``\ ing the Task or Trigger that fired first.
```

Below we see it used to implement a timeout.

.. code-block:: python3

    @cocotb.test

>     async def test_thingy(dut):

Reviewer: ```suggestion
    async def test_quiesce_or_timeout(dut):
```
Easier to point users to the right code example

        # generate stimulus and drive it to the design
        for trans in generate_transactions():
            await drive(dut.intf, trans)

        # wait for the design to quiesce or timeout

>         timeout = Timer(10, 'us')

Reviewer: ```suggestion
        timeout = Timer(10, "us")
```
Use black-like quotes

        res = await First(timeout, quiesce())
        assert res is not timeout, "Design has hung!"

Fortunately for users timeouts are a common operation and cocotb provides :func:`~cocotb.triggers.with_timeout`.

> The second section in the above code using it would be ``await with_timeout(quiesce(), 10, 'us')``.

Reviewer: ```suggestion
The second section in the above code using it would be ``await with_timeout(quiesce(), 10, "us")``.
```

.. note::

    :class:`~cocotb.triggers.First` does *not* cancel Tasks that did not complete after it returns.
    This means that Tasks passed to it are *still running*.
    You may need to cancel those Tasks with :meth:`.Task.cancel`.

Additionally, :class:`~cocotb.triggers.First` can be used to determine which of multiple Tasks complete first using the following idiom.

.. code-block:: python3

    @cocotb.test

>     async def test_thingy(dut):

Reviewer: ```suggestion
    async def test_which_finished_first(dut):
```

        task_A = cocotb.start_soon(drive_A())
        task_B = cocotb.start_soon(drive_B())

        res = await First(task_A.complete, task_B.complete)
        if res is task_A.complete:
            cocotb.log.info("Input A finished first")
        else:
            cocotb.log.info("Input B finished first")

> :class:`~cocotb.triggers.Combine` is like ``await``\ ing multiple Triggers or Tasks at the same time,

Reviewer: ```suggestion
Waiting Until Multiple Events Have Happened
-------------------------------------------

:class:`~cocotb.triggers.Combine` is like ``await``\ ing multiple Triggers or Tasks at the same time,
```

but it resumes after *all* the listed Triggers or Tasks fire.

> Using the example from before, we can use it to wait until both the driving and quiesce are done.

Reviewer: ```suggestion
Using the example from the previous section, we can use it to wait until both the driving and quiesce are done.
```

.. code-block:: python3

    @cocotb.test

>     async def test_thingy(dut):

Reviewer: ```suggestion
    async def test_wait_for_both(dut):
```

        # generate stimulus and drive it to the design
        async def drive_transactions():
            for trans in generate_transactions():
                await drive(dut.intf, trans)

        # wait for both the driving and quiescing to complete before continuing
        await Combine(drive_transactions(), quiesce())

And of course, the sky is the limit when you compose the two.

.. code-block:: python3

    @cocotb.test

>     async def test_thingy(dut):

Reviewer: ```suggestion
    async def test_wait_for_both_with_timeout(dut):
```

        # wait for both the driving and quiescing to complete before continuing
        # but timeout if *either* the driving or settling take too long
        await Combine(

>             with_timeout(drive_transactions(), 1, 'us'),
>             with_timeout(quiesce(), 10, 'us'),

Reviewer: ```suggestion
            with_timeout(drive_transactions(), 1, "us"),
            with_timeout(quiesce(), 10, "us"),
```
